crates/ruff_linter/src/rules/pylint/rules/unnecessary_dunder_call.rs

```diff
@@ -15,6 +15,12 @@ use ruff_python_ast::PythonVersion;
 /// Dunder names are not meant to be called explicitly and, in most cases, can
 /// be replaced with builtins or operators.
 ///
+/// ## Fix safety
+/// This fix is sometimes unsafe. When replacing dunder method calls with operators
+/// or builtins, the fix must carefully consider operator precedence to avoid changing
+/// the code's behavior. For example, without proper handling, changing `(-a).__sub__(b)`
+/// to `-a - b` would change the execution order due to operator precedence rules.
```

Comment:
Like the other rule I reviewed, I think this fix is also always unsafe. Otherwise I think this explanation is pretty good. I think I would change it slightly to emphasize the custom dunder methods more heavily (as in [this comment](https://github.com/astral-sh/ruff/issues/9572#issuecomment-1900031398)). I think we handle the operator precedence correctly (barring bugs in the implementation, which should be fixed, not make a rule unsafe!), so the unsafety is from the rule not resolving dunder methods, at least based on my reading. Maybe something like this:

```suggestion
/// This fix is marked as unsafe because replacing dunder method calls with operators
/// or builtins may change the code's behavior in cases where a type only implements a
/// subset of the related methods, e.g. `__radd__`  but not `__add__`.
```

I'm not totally happy with that, but hopefully it pushes us in the right direction.

Reply:
Here are two more ways the fix is unsafe because it can change behavior:
```console
$ cat >plc2801.py <<'# EOF'
class C: pass
c = C()
c.__str__ = lambda: "c"
print(c.__str__().split()[0])
print(c.__gt__(1))
# EOF

$ python plc2801.py
c
NotImplemented

$ ruff --isolated check plc2801.py --select PLC2801 --preview --unsafe-fixes --fix
Found 2 errors (2 fixed, 0 remaining).

$ python plc2801.py
<__main__.C
Traceback (most recent call last):
  File "plc2801.py", line 3, in <module>
    print(c > 1)
          ^^^^^^^
TypeError: '>' not supported between instances of 'C' and 'int'

Reply:
Nice, thanks. I find it so weird that these print `NotImplemented` instead of raising some kind of exception. Maybe we can summarize this as something like "... may change the program's behavior for types with custom dunder method implementations."

This isn't really relevant for this PR, but I guess we could potentially mark the fix safe if we could infer that the types are built-in? Correspondingly, we should not offer a fix if we know the dunder method has been changed, as in these examples, but I think that requires a lot more type inference in the general case, a la ty (formerly red-knot).

Reply:
It can also change the program’s behavior for types _without_ custom dunder method implementations. In my example above, there are no custom `__gt__` implementations, and yet the fix for `__gt__` changes behavior. The fix is unsafe whenever the original code doesn’t do exactly what the replacement does, which is vague, but there are so many subtleties it’s hard to write a succinct yet helpful summary.

It can still be unsafe with built-in types:
```pycon
>>> (1).__gt__(1.0)
NotImplemented
>>> 1 > 1.0
False
```
But yes, you’re right that there are some combinations of built-in types and dunder methods that are safe to fix.

Reply:
Ah of course, thanks for the corrections. I was too eager to find a short description. We may just want to mention a few of these examples.